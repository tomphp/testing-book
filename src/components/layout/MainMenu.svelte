<script context="module" lang="ts">
	import type { SubItemProps } from './MainMenu/Item.svelte';

	export interface ItemProps {
		title: string;
		link?: string;
		children?: SubItemProps[];
	}
</script>

<script lang="ts">
	import MainMenuItem from './MainMenu/Item.svelte';

	export let items: ItemProps[];
</script>

<nav role="menu" class="pl-5">
	<ul class="lg:flex">
		{#each items as { title, link, children }}
			<MainMenuItem {title} {link} {children} />
		{/each}
	</ul>
</nav>
