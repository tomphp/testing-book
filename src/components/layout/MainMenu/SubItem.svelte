<script lang="ts">
  export let link: string;
  export let title: string;
</script>

<li>
  <a
    class="
      block
      w-full
      whitespace-nowrap
      bg-transparent
      py-2
      px-4
      text-sm
      font-normal
      text-neutral-700
      hover:bg-neutral-100
      active:text-neutral-800
      active:no-underline
      disabled:pointer-events-none
      disabled:bg-transparent
      disabled:text-neutral-400"
    href="{link}"
    data-te-dropdown-item-ref>{title}</a>
</li>
