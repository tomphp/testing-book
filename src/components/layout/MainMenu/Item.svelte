<script lang="ts" context="module">
	export interface SubItemProps {
		title: string;
		link: string;
	}
</script>

<script lang="ts">
	import ArrowDown from './ArrowDown.svelte';
	import SubItem from './SubItem.svelte';

	export let link: string | undefined = undefined;
	export let title: string;
	export let children: SubItemProps[] | undefined = undefined;

	let open = false;

	function openSubmenu(): void {
		open = true;
	}

	function closeSubmenu(): void {
		open = false;
	}

	$: href = link;
</script>

<li class="font-bold text-center p-2">
	<div class="justify-center">
		<div on:mouseenter={() => openSubmenu()} on:mouseleave={() => closeSubmenu()}>
			<div class="relative" data-te-dropdown-ref>
				<a
					class="
              flex
              text-sm
              items-center
              whitespace-nowrap
              rounded
              bg-primary
              lg:px-6
              pt-2.5
              pb-2
              font-medium
              uppercase
              leading-normal
              text-white
              shadow-[0_4px_9px_-4px_#3b71ca]
              transition
              duration-150
              ease-in-out
              hover:bg-white
              hover:text-emerald-500
              hover:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)]
              focus:bg-primary-600
              focus:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)]
              focus:outline-none
              focus:ring-0
              active:bg-primary-700
              active:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)]
              motion-reduce:transition-none"
					{href}
					type="button"
					id="dropdownMenuButton2"
					data-te-dropdown-toggle-ref
					aria-expanded="false"
					data-te-ripple-init
					data-te-ripple-color="light"
				>
					{title}
					{#if children}
						<span class="ml-2 w-2"><ArrowDown /></span>
					{/if}
				</a>
				{#if children}
					<ul
						class:hidden={!open}
						class="
                absolute
                z-[1000]
                float-left
                m-0
                min-w-max
                list-none
                overflow-hidden
                rounded-lg
                border-none
                bg-white
                bg-clip-padding
                text-left
                text-base
                shadow-lg
                [&[data-te-dropdown-show]]:block"
						aria-labelledby="dropdownMenuButton2"
						data-te-dropdown-menu-ref
					>
						{#each children as { link, title }}
							<SubItem {link} {title} />
						{/each}
					</ul>
				{/if}
			</div>
		</div>
	</div>
</li>
