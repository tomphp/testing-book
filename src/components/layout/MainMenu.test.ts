import {test, expect} from 'vitest';
import '@testing-library/jest-dom';
import {render, screen} from "@testing-library/svelte";
import MainMenu from "./MainMenu.svelte";

const testItems = [
  {title: "No Link or Children"},
  {title: "Link Without Children", link: "/foo"},
  {title: "Link WithChildren", link: "/bar", children: [
    { link: "/bar/1", title: "Child 1"},
    { link: "/bar/2", title: "Child 2"}
  ]},
];

test("MainMenu snapshot", () => {
  const {container} = render(MainMenu, {props: {items: testItems}});
  expect(container).toMatchSnapshot();
});
