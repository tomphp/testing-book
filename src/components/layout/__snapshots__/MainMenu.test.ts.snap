// Vitest Snapshot v1

exports[`MainMenu snapshot 1`] = `
<body>
  <div>
    <nav
      class="pl-5"
      role="menu"
    >
      <ul
        class="flex"
      >
        <li
          class="mr-0 font-bold text-center p-2"
        >
          <div
            class="flex justify-center"
          >
            <div>
              <div
                class="relative"
                data-te-dropdown-ref=""
              >
                <a
                  aria-expanded="false"
                  class="flex text-sm items-center whitespace-nowrap rounded bg-primary px-6 pt-2.5 pb-2 font-medium uppercase leading-normal text-white shadow-[0_4px_9px_-4px_#3b71ca] transition duration-150 ease-in-out hover:bg-white hover:text-emerald-500 hover:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:bg-primary-600 focus:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:outline-none focus:ring-0 active:bg-primary-700 active:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] motion-reduce:transition-none"
                  data-te-dropdown-toggle-ref=""
                  data-te-ripple-color="light"
                  data-te-ripple-init=""
                  id="dropdownMenuButton2"
                  type="button"
                >
                  No Link or Children
                   
                </a>
                 
              </div>
            </div>
          </div>
        </li>
        <!--&lt;Item&gt;-->
        <li
          class="mr-0 font-bold text-center p-2"
        >
          <div
            class="flex justify-center"
          >
            <div>
              <div
                class="relative"
                data-te-dropdown-ref=""
              >
                <a
                  aria-expanded="false"
                  class="flex text-sm items-center whitespace-nowrap rounded bg-primary px-6 pt-2.5 pb-2 font-medium uppercase leading-normal text-white shadow-[0_4px_9px_-4px_#3b71ca] transition duration-150 ease-in-out hover:bg-white hover:text-emerald-500 hover:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:bg-primary-600 focus:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:outline-none focus:ring-0 active:bg-primary-700 active:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] motion-reduce:transition-none"
                  data-te-dropdown-toggle-ref=""
                  data-te-ripple-color="light"
                  data-te-ripple-init=""
                  href="/foo"
                  id="dropdownMenuButton2"
                  type="button"
                >
                  Link Without Children
                   
                </a>
                 
              </div>
            </div>
          </div>
        </li>
        <!--&lt;Item&gt;-->
        <li
          class="mr-0 font-bold text-center p-2"
        >
          <div
            class="flex justify-center"
          >
            <div>
              <div
                class="relative"
                data-te-dropdown-ref=""
              >
                <a
                  aria-expanded="false"
                  class="flex text-sm items-center whitespace-nowrap rounded bg-primary px-6 pt-2.5 pb-2 font-medium uppercase leading-normal text-white shadow-[0_4px_9px_-4px_#3b71ca] transition duration-150 ease-in-out hover:bg-white hover:text-emerald-500 hover:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:bg-primary-600 focus:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] focus:outline-none focus:ring-0 active:bg-primary-700 active:shadow-[0_8px_9px_-4px_rgba(59,113,202,0.3),0_4px_18px_0_rgba(59,113,202,0.2)] motion-reduce:transition-none"
                  data-te-dropdown-toggle-ref=""
                  data-te-ripple-color="light"
                  data-te-ripple-init=""
                  href="/bar"
                  id="dropdownMenuButton2"
                  type="button"
                >
                  Link WithChildren
                   
                  <span
                    class="ml-2 w-2"
                  >
                    <svg
                      class="h-5 w-5"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
                        fill-rule="evenodd"
                      />
                    </svg>
                    <!--&lt;ArrowDown&gt;-->
                  </span>
                </a>
                 
                <ul
                  aria-labelledby="dropdownMenuButton2"
                  class="absolute z-[1000] float-left m-0 min-w-max list-none overflow-hidden rounded-lg border-none bg-white bg-clip-padding text-left text-base shadow-lg [&[data-te-dropdown-show]]:block hidden"
                  data-te-dropdown-menu-ref=""
                >
                  <li>
                    <a
                      class="block w-full whitespace-nowrap bg-transparent py-2 px-4 text-sm font-normal text-neutral-700 hover:bg-neutral-100 active:text-neutral-800 active:no-underline disabled:pointer-events-none disabled:bg-transparent disabled:text-neutral-400"
                      data-te-dropdown-item-ref=""
                      href="/bar/1"
                    >
                      Child 1
                    </a>
                  </li>
                  <!--&lt;SubItem&gt;-->
                  <li>
                    <a
                      class="block w-full whitespace-nowrap bg-transparent py-2 px-4 text-sm font-normal text-neutral-700 hover:bg-neutral-100 active:text-neutral-800 active:no-underline disabled:pointer-events-none disabled:bg-transparent disabled:text-neutral-400"
                      data-te-dropdown-item-ref=""
                      href="/bar/2"
                    >
                      Child 2
                    </a>
                  </li>
                  <!--&lt;SubItem&gt;-->
                </ul>
              </div>
            </div>
          </div>
        </li>
        <!--&lt;Item&gt;-->
      </ul>
    </nav>
    <!--&lt;MainMenu&gt;-->
  </div>
</body>
`;
